### PR TITLE
Implement manifest selection

### LINKED ISSUE #17: Watch manifest & update state accordingly
In case the user modified the manifest to add a dependency, we should automatically offer to re-build or auto-update the config at least.

### PR DESCRIPTION
**Pull request by SeaDve**
Closes #17 and #8

TODO:
- [x] cleaner implementation and cleanup
- [x] handle new modules addition changes on manifest
- [x] watch manifest for on disk changes
- [x] status bar item for current active manifest

### CODE REVIEW

File: src/flatpakManifest.ts [hunk 1]
```diff
@@ -467,6 +468,12 @@ export class FlatpakManifest {
         return new Command('flatpak', args, this.workspace)
     }
 
+    async deleteBuildDir(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifest.ts:471]: and it complains when tried to build init when .flatpak exists. I haven't tested much the edge cases yet, maybe tomorrow

**Comment by SeaDve** [comment]
Most of the things are now implemented and probably need a commit squash.

One thing I'm not sure is whether the status bar item should open the active manifest or activate `select-manifest` command.

It also needs more testing for edge cases

File: src/flatpakManifestManager.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            console.log(`Possible manifest created at ${newUri.fsPath}`)
+            const newManifest = await this.finder.parseManifest(newUri)
+            if (newManifest !== null) {
+                this.manifests?.add(newManifest)
+            }
+        })
+        this.manifestWatcher.onDidChange(async (uri) => {
+            console.log(`Possible manifest modified at ${uri.fsPath}`)
+            const activeManifest = this.getActiveManifest()
+            if (uri.fsPath !== activeManifest?.uri.fsPath) {
```
Comment (SeaDve) [inline src/flatpakManifestManager.ts:43]: Do we want to also update the non active manifests too?

Comment (SeaDve) [reply src/flatpakManifestManager.ts:47]: Currently, it may produce unexpected behavior when the user modified an unactive manifest then switch to it.

File: src/flatpakManifest.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifest.ts:481]: hmm maybe just the `.flatpak/repo`

File: package.json [hunk 3]
```diff
@@ -202,7 +207,6 @@
   "dependencies": {
     "@types/js-yaml": "^4.0.5",
     "dbus-next": "^0.10.2",
-    "effector": "^22.2.0",
```
Comment (bilelmoussaoui) [inline package.json:205]: nice

File: src/workspaceState.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import * as vscode from 'vscode'
+
+type WorkspaceStateKey = 'LastActiveManifestUri' | 'Initialized' | 'DependenciesUpdated' | 'DependenciesBuilt' | 'ApplicationBuilt'
+
+export class WorkspaceState {
```
Comment (bilelmoussaoui) [inline src/workspaceState.ts:5]: Would be nice to have small comments to explain what is this used for, just to make it easier for potential future contributors

File: src/workspaceState.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    async resetPipelineState(): Promise<void> {
+        await this.setInitialized(false)
+        await this.setDependenciesUpdated(false)
+        await this.setDependenciesBuilt(false)
+        await this.setApplicationBuilt(false)
+    }
+
+    async setContext(contextName: string, value: boolean): Promise<void> {
```
Comment (bilelmoussaoui) [inline src/workspaceState.ts:19]: Should this be private?

File: src/workspaceState.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
+import * as vscode from 'vscode'
+
+type WorkspaceStateKey = 'LastActiveManifestUri' | 'Initialized' | 'DependenciesUpdated' | 'DependenciesBuilt' | 'ApplicationBuilt'
```
Comment (bilelmoussaoui) [inline src/workspaceState.ts:3]:
```suggestion
type WorkspaceStateKey = 'ActiveManifestUri' | 'Initialized' | 'DependenciesUpdated' | 'DependenciesBuilt' | 'ApplicationBuilt'
```

File: src/flatpakRunnerStatusItem.ts [hunk 7]
```diff
@@ -12,12 +12,12 @@ export interface Status {
     clickable: Clickable | null
 }
 
-export class StatusBarItem {
+export class FlatpakRunnerStatusItem extends vscode.Disposable {
     private inner: vscode.StatusBarItem
 
-    constructor(extCtx: vscode.ExtensionContext) {
+    constructor() {
+        super(() => this.dispose())
```
Comment (bilelmoussaoui) [inline src/flatpakRunnerStatusItem.ts:19]: Not sure I understand the needs for calling dispose on the constructor, is this to make sure the old instance is properly disposed?

File: src/flatpakManifest.ts [hunk 8]
```diff
@@ -27,7 +27,26 @@ export class FlatpakManifest {
...
         this.stateDir = path.join(this.buildDir, 'flatpak-builder')
-        this.stateFile = path.join(this.buildDir, 'pipeline.json')
+    }
+
+    /**
+     * Update the fields of this
+     * @param other the more updated FlatpakManifest to get new info from
+     * @returns whether to request for a rebuild
+     */
+    updateFieldsFrom(other: FlatpakManifest): boolean {
```
Comment (bilelmoussaoui) [inline src/flatpakManifest.ts:37]: Do we need to update the fields here? normally we would have separate FlatpakManifest or not?

**Review by bilelmoussaoui — COMMENTED** [review]
I did a quick review, I will get back to this later today

File: src/flatpakRunnerStatusItem.ts (hunk 7 above)
Comment (SeaDve) [reply src/flatpakRunnerStatusItem.ts:19]: Its a requirement when extending Disposable. According to vscode API docs, implementing Diposable and pushing it in subscriptions automatically free the resources when the extension deactivates. We could manually push each disposable (like the StatusBarItem) to the subscriptions, but we need to pass ExtensionContext everywhere, which I think is not any better.

File: src/flatpakManifest.ts (hunk 8 above)
Comment (SeaDve) [reply src/flatpakManifest.ts:37]: Either could work, but the intent is shown better and clearer this way.

Comment (bilelmoussaoui) [reply src/flatpakManifest.ts:37]: I just noticed you are calling this when the manifest receives changes, I think the best way to handle this is to just the new modified manifest as the active one (if it has the same uri as the currently active one)

File: src/workspaceState.ts (hunk 5 above)
Comment (SeaDve) [reply src/workspaceState.ts:19]: this is also used by manifestManager when a manifest is found

Comment (SeaDve) [reply src/workspaceState.ts:19]: ahh, we can just update the context here and make it private

File: src/flatpakManifest.ts (hunk 8 above)
Comment (SeaDve) [reply src/flatpakManifest.ts:37]: changed to that behavior, and since an `activeManifestChanged` triggers to clear `repoDir`. It will ask to initialize a build every manifest modification. So I guess we can remove that message and just handle the build initialization automatically on a build or rebuild command.

Comment (bilelmoussaoui) [reply src/flatpakManifest.ts:37]: > It will ask to initialize a build every manifest modification. 

But normally we already know if the manifest uri changed, if not and that manifest state says it was initialized that should give us enough info to know whether we should show the dialog or not

Comment (SeaDve) [reply src/flatpakManifest.ts:37]: Nvm, that was not the reason. `===` operator in js doesn't check by value, but by instance. So had to use `util.isDeepStrictEqual`,  either way, I think it is still better to remove that message, since running any build command should initialize a build

Comment (SeaDve) [reply src/flatpakManifest.ts]: Oops, sorry. I didn't notice your last message. Was it okay though to remove the message with this reason?


> running any build command should initialize a build

Comment (bilelmoussaoui) [reply src/flatpakManifest.ts]: >  running any build command should initialize a build

It should, but my intention with that dialog is that someone not familiar with Flatpak/VScode could get an idea how to start a build and get the application up and running easily

**Comment by bilelmoussaoui** [comment]
Btw, I sent you an invite for having write access to the repository just in case :)

File: src/extension.ts [hunk 9]
```diff
@@ -1,193 +1,330 @@
...
+    if (manifest === null) {
+      return
+    }
+
+    if (!isLastActive) {
+      this.runner.close()
+      await this.resetPipelineState()
+
+      this.terminalProvider?.dispose()
+      this.terminalProvider === undefined
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/extension.ts:222]: oops, It should have been `this.terminalProvider === undefined`

File: src/extension.ts [hunk 10]
```diff
@@ -1,193 +1,330 @@
...
+    }
+  }
+
+  private async handleFinishedTask(finishedTask: FinishedTask): Promise<void> {
+    switch (finishedTask.mode) {
+      case TaskMode.buildInit: {
+        await this.workspaceState.setInitialized(true)
+        const activeManifest = this.manifestManager.getActiveManifest()
+        if (activeManifest) {
+          void this.loadIntegrations(activeManifest)
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/extension.ts:265]: So it doesn't have to await the promise. Could use the `.then(() => {}, () => {}), but this is cleaner. https://typescript-eslint.io/rules/no-floating-promises/#ignorevoid

[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply src/extension.ts:265]: Oh good to know

**Comment by SeaDve** [comment]
thanks :D

File: src/extension.ts [hunk 11]
```diff
@@ -1,193 +1,330 @@
...
     })
 
-    context.subscriptions.push(
-      registerCommand(`${EXT_ID}.runtime-terminal`, () => {
-        const runtimeTerminal = window.createTerminal(manifest.runtimeTerminal())
+    this.registerCommand('runtime-terminal', async () => {
+      await this.manifestManager.doWithActiveManifestElseDisplayErr((activeManifest) => {
+        const runtimeTerminal = window.createTerminal(activeManifest.runtimeTerminal())
+        this.extCtx.subscriptions.push(runtimeTerminal)
+        this.activeTerminals.push(runtimeTerminal)
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/extension.ts:73]: ah yeah, that seems problematic. It is only used though to kill terminals when manifest changes. I think we can remove that since we can't also kill terminals that are provided by TerminalProviders.

File: README.md [hunk 12]
```diff
@@ -26,6 +26,7 @@ A very simple VSCode extension that detects a Flatpak manifest and offers variou
 - Runtime Terminal: Spawn a new terminal inside the specified SDK.
 - Build Terminal: Spawn a new terminal inside the current build repository (Note that the SDKs used are automatically mounted and enabled as well).
 - Show Output Terminal: Show the output terminal of the build and run commands.
+- Select Manifest: Change the active manifest.
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply README.md:29]: sure

**Comment by bilelmoussaoui** [comment]
Could use a rebase

**Comment by SeaDve** [comment]
I'll squash every commits for easier rebase

**Comment by SeaDve** [comment]
image

It shows this message when there is some error on the manifest

File: src/extension.ts [hunk 13]
```diff
@@ -1,206 +1,345 @@
...
+      await this.resetPipelineState()
+
+      this.terminalProvider?.dispose()
+      this.terminalProvider = undefined
+
+      for (const terminal of this.activeTerminals) {
+        terminal.dispose()
+      }
+
+      await manifest.deleteRepoDir()
```
Comment (bilelmoussaoui) [inline src/extension.ts:240]: We should also disable the integrations here as well

File: src/extension.ts [hunk 14]
```diff
@@ -1,206 +1,345 @@
...
+    }
+
+    this.terminalProvider = new TerminalProvider(manifest)
+
+    await this.ensurePipelineState()
+  }
+
+  private async resetPipelineState(): Promise<void> {
```
Comment (bilelmoussaoui) [inline src/extension.ts:248]: Should probably be part of workspaceState and called reset :)

File: src/extension.ts [hunk 15]
```diff
@@ -1,206 +1,345 @@
...
+        }
+        if (!finishedTask.restore) {
+          void this.executeCommand(TaskMode.updateDeps, finishedTask.completeBuild)
+        }
+        break
+      }
+      case TaskMode.updateDeps:
+        await this.workspaceState.setDependenciesUpdated(true)
+        await this.workspaceState.setDependenciesBuilt(false)
+        // Assume user might want to rebuild dependencies
```
Comment (bilelmoussaoui) [inline src/extension.ts:287]: This comment  should be moved to the line before

File: src/flatpakManifestManager.ts [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+    label: string,
+    detail: string,
+    manifest: FlatpakManifest
+}
+
+export class FlatpakManifestManager implements vscode.Disposable {
+    private readonly statusItem: vscode.StatusBarItem
+    private readonly workspaceState: WorkspaceState
+    private readonly manifestWatcher: vscode.FileSystemWatcher
+    private manifests?: FlatpakManifestMap
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestManager.ts:21]: it's for consistency with flatpakManifest. We should rename FlatpakManifest too I guess to match

File: src/flatpakRunner.ts [hunk 17]
```diff
@@ -1,11 +1,24 @@
 import * as vscode from 'vscode'
-import { failure, finished, newTask } from './store'
-import { TaskMode } from './taskMode'
+import { TaskMode, taskModeAsStatus } from './taskMode'
 import { Command } from './command'
 import * as pty from './nodePty'
 import { FlatpakTerminal } from './flatpakTerminal'
+import { FlatpakRunnerStatusItem } from './flatpakRunnerStatusItem'
+import { EXT_ID } from './extension'
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakRunner.ts:7]: we could maybe handle the bunch of renames on #84

File: src/flatpakRunner.ts [hunk 18]
```diff
@@ -1,11 +1,24 @@
 import * as vscode from 'vscode'
-import { failure, finished, newTask } from './store'
-import { TaskMode } from './taskMode'
+import { TaskMode, taskModeAsStatus } from './taskMode'
 import { Command } from './command'
 import * as pty from './nodePty'
 import { FlatpakTerminal } from './flatpakTerminal'
+import { FlatpakRunnerStatusItem } from './flatpakRunnerStatusItem'
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakRunner.ts:6]: let's handle this too ij #84

File: src/extension.ts (hunk 14 above)
Comment (SeaDve) [reply src/extension.ts:248]: we dont want to reset activeManifestUri too. I removed it in the workspaceState as workspaceState shoudn't know which keys are part of the pipeline state

File: src/flatpakManifestMap.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
+import { FlatpakManifest } from './flatpakManifest'
+import * as vscode from 'vscode'
+
+export class FlatpakManifestMap {
+    private readonly inner: Map<string, FlatpakManifest>
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestMap.ts:5]: The readonly is to prevent redeclaring the map other than in the constructor

File: src/flatpakManifestManager.ts [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+import { FlatpakManifestMap } from './flatpakManifestMap'
+import { findManifests, getManifestPathGlobPattern, parseManifest } from './flatpakManifestUtils'
+
+export interface ManifestQuickPickItem {
+    label: string,
+    detail: string,
+    manifest: FlatpakManifest
+}
+
+export class FlatpakManifestManager implements vscode.Disposable {
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestManager.ts:17]: let's handle this too in #84

File: src/flatpakManifestManager.ts [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+        let lastActiveManifestUri = this.workspaceState.getActiveManifestUri()
+
+        if (lastActiveManifestUri !== undefined && !await exists(lastActiveManifestUri.fsPath)) {
+            lastActiveManifestUri = undefined
+        }
+
+        const manifests = await this.getManifests()
+
+        if (lastActiveManifestUri === undefined) {
+            if (manifests.size() !== 1) {
```
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply src/flatpakManifestManager.ts:78]: Could use a comment then because it wasn't clear from going through the code

File: src/flatpakManifestManager.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestManager.ts:21]: Will handle it in #84

File: src/flatpakManifestManager.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestManager.ts]: Added a comment

File: src/flatpakManifestManager.ts [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+    readonly onDidActiveManifestChanged = this._onDidActiveManifestChanged.event
+
+    private readonly _onDidRequestRebuild = new vscode.EventEmitter<FlatpakManifest>()
+    readonly onDidRequestRebuild = this._onDidRequestRebuild.event
+
+    constructor(workspaceState: WorkspaceState) {
+        this.workspaceState = workspaceState
+        this.activeManifest = null
+
+        this.manifestWatcher = vscode.workspace.createFileSystemWatcher(getManifestPathGlobPattern())
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestManager.ts:34]: yeah, that should have been a const

File: src/flatpakManifestManager.ts [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+        await this.setActiveManifest(selected.manifest, false)
+        return selected.manifest
+    }
+
+    /**
+     * Convenience function to do things with the active manifest if it exist.
+     * If it doesn't exist. Show manifests picker, or show messages.
+     * @param func Callback function where the active manifest can be handled
+     */
+    async doWithActiveManifest(func: (manifest: FlatpakManifest) => Promise<void> | void): Promise<void> {
```
[the remark this answers is not in the document]
Comment (SeaDve) [reply src/flatpakManifestManager.ts:216]: The `doWithActiveManifest` also acts like a wrapper function so if the callback func is async, the function has also be async
